File: src/utils/matlab/libs/gps_l1_ca_read_telemetry_dump.m
% -------------------------------------------------------------------------
%
% GNSS-SDR is a Global Navigation Satellite System software-defined receiver.
% This file is part of GNSS-SDR.
%
% SPDX-FileCopyrightText: Javier Arribas 2011
% SPDX-License-Identifier: GPL-3.0-or-later
%
% -------------------------------------------------------------------------

function [telemetry] = gps_l1_ca_read_telemetry_dump (filename, count)

%%
%% open GNSS-SDR tracking binary log file .dat and return the contents
%%

m = nargchk (1,2,nargin);
num_double_vars=3;
double_size_bytes=8;
num_int_vars=2;
int_size_bytes=4;
skip_bytes_each_read=double_size_bytes*num_double_vars+num_int_vars*int_size_bytes;
bytes_shift=0;
if (m)
    usage (m);
end

if (nargin < 3)
    count = Inf;
end
f = fopen (filename, 'rb');
if (f < 0)
else
    [x, loops_counter] = fread (f,skip_bytes_each_read);
    fseek(f,0,-1);
    for i=1:min(count, loops_counter),
        telemetry(i).tow_current_symbol_ms = fread (f, count, 'float64',skip_bytes_each_read-double_size_bytes);
        bytes_shift=bytes_shift+double_size_bytes;
        fseek(f,bytes_shift,'bof'); % move to next interleaved
        telemetry(i).tracking_sample_counter = fread (f, count, 'uint64',skip_bytes_each_read-double_size_bytes);
        bytes_shift=bytes_shift+double_size_bytes;
        fseek(f,bytes_shift,'bof'); % move to next interleaved
        telemetry(i).tow = fread (f, count, 'float64',skip_bytes_each_read-double_size_bytes);
        bytes_shift=bytes_shift+double_size_bytes;
        fseek(f,bytes_shift,'bof'); % move to next interleaved
        telemetry(i).nav_simbols = fread (f, count, 'int32',skip_bytes_each_read-int_size_bytes);
        bytes_shift=bytes_shift+int_size_bytes;
        fseek(f,bytes_shift,'bof'); % move to next interleaved
        telemetry(i).prn = fread (f, count, 'int32',skip_bytes_each_read-int_size_bytes);
        bytes_shift=bytes_shift+int_size_bytes;
        fseek(f,bytes_shift,'bof'); % move to next interleaved
    end
    fclose (f);
end

File: src/utils/python/lib/gps_l1_ca_read_telemetry_dump.py
"""
 gps_l1_ca_read_telemetry_dump.py
   gps_l1_ca_read_telemetry_dump (filename)

 Open and read GNSS-SDR telemetry binary log files (.dat) into Python, and
 return the contents.

 Irene Pérez Riega, 2023. iperrie@inta.es

      Args:
        filename        - Path to file telemetry[N].dat with the raw data

      Return:
        telemetry       - A dictionary with the processed data

 -----------------------------------------------------------------------------

 GNSS-SDR is a Global Navigation Satellite System software-defined receiver.
 This file is part of GNSS-SDR.

 Copyright (C) 2022  (see AUTHORS file for a list of contributors)
 SPDX-License-Identifier: GPL-3.0-or-later

 -----------------------------------------------------------------------------
"""

import struct


def gps_l1_ca_read_telemetry_dump(filename):

    double_size_bytes = 8
    int_size_bytes = 4
    bytes_shift = 0

    tow_current_symbol_ms = []
    tracking_sample_counter = []
    tow = []
    nav_simbols = []
    prn = []

    f = open(filename, 'rb')
    if f is None:
        return None
    else:
        while True:
            f.seek(bytes_shift, 0)

            tow_current_symbol_ms.append(struct.unpack(
                'd', f.read(double_size_bytes))[0])
            bytes_shift += double_size_bytes
            f.seek(bytes_shift, 0)

            tracking_sample_counter.append(struct.unpack(
                'Q', f.read(double_size_bytes))[0])
            bytes_shift += double_size_bytes
            f.seek(bytes_shift, 0)

            tow.append(struct.unpack(
                'd', f.read(double_size_bytes))[0])
            bytes_shift += double_size_bytes
            f.seek(bytes_shift, 0)

            nav_simbols.append(struct.unpack(
                'I', f.read(int_size_bytes))[0])
            bytes_shift += int_size_bytes
            f.seek(bytes_shift, 0)

            prn.append(struct.unpack('I', f.read(int_size_bytes))[0])
            bytes_shift += int_size_bytes
            f.seek(bytes_shift, 0)

            # Check file
            linea = f.readline()
            if not linea:
                break

        telemetry = {
            'tow_current_symbol_ms': tow_current_symbol_ms,
            'tracking_sample_counter': tracking_sample_counter,
            'tow': tow,
            'nav_simbols': nav_simbols,
            'prn': prn
        }

    return telemetry
